sanic/websocket.py

```diff
@@ -7,23 +10,748 @@
+        if no message is received within ``timeout`` seconds, return ``None``.
+        Set ``timeout`` to ``0`` to check if a message was already received.
+        :raises ~websockets.exceptions.ConnectionClosed: when the
+            connection is closed
+        :raises RuntimeError: if two tasks call :meth:`recv` or
+            :meth:`recv_streaming` concurrently
+        """
+
+        # TODO HANDLE THE SITUATION WHERE THE CONNECTION IS CLOSED
+
+        if self.recv_lock.locked():
+            raise RuntimeError(
+                "cannot call recv while another task "
+                "is already waiting for the next message"
+            )
+        await self.recv_lock.acquire()
+        try:
+            return await self.assembler.get(timeout)
+        finally:
+            self.recv_lock.release()
+
+    async def recv_burst(self, max_recv=256) -> Sequence[Data]:
+        """
+        Receive the messages which have arrived since last checking.
+        Return a :class:`list` containing :class:`str` for a text frame
+        and :class:`bytes` for a binary frame.
+        When the end of the message stream is reached, :meth:`recv_burst`
+        raises :exc:`~websockets.exceptions.ConnectionClosed`. Specifically,
+        it raises :exc:`~websockets.exceptions.ConnectionClosedOK` after a
+        normal connection closure and
+        :exc:`~websockets.exceptions.ConnectionClosedError` after a protocol
+        error or a network failure.
+        :raises ~websockets.exceptions.ConnectionClosed: when the
+            connection is closed
+        :raises RuntimeError: if two threads call :meth:`recv` or
+            :meth:`recv_streaming` concurrently
+        """
+
+        # TODO HANDLE THE SITUATION WHERE THE CONNECTION IS CLOSED
+
+        if self.recv_lock.locked():
+            raise RuntimeError(
+                "cannot call recv_burst while another task "
+                "is already waiting for the next message"
+            )
+        await self.recv_lock.acquire()
+
+        messages = []
+        try:
+            # Prevent pausing the transport when we're
+            # receiving a burst of messages
+            self.can_pause = False
+            while True:
+                 m = await self.assembler.get(timeout=0)
+                 if m is None:
+                     # None left in the burst. This is good!
+                     break
+                 messages.append(m)
+                 if len(messages) >= max_recv:
+                     # Too much data in the pipe. Hit our burst limit.
+                     break
+                 # Allow an eventloop iteration for the
+                 # next message to pass into the Assembler
+                 await asyncio.sleep(0)
+        finally:
+            self.can_pause = True
+            self.recv_lock.release()
+        return messages
+
+    async def recv_streaming(self) -> AsyncIterator[Data]:
```

Comment:
Pretty cool — I never implemented this [feature request](https://github.com/aaugustin/websockets/issues/479) in websockets itself!

Reply:
It was in your sample implementation, so I've put it into our implementation. I've tested it, seems to work fine, and I'm sure our users will take advantage of it for streaming in large messages.

Reply:
Me:

![Dory](https://static.wikia.nocookie.net/pixar/images/1/1f/Dory-white.jpg/revision/latest/scale-to-width-down/620?cb=20110924203518)

Reply:
![image](https://user-images.githubusercontent.com/166269/122217350-0220b600-ceb6-11eb-9d79-fe52f2afcfd2.png)